hazelcast/src/main/java/com/hazelcast/cache/impl/journal/RingbufferCacheEventJournalImpl.java

```diff
@@ -0,0 +1,256 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl.journal;
+
+import com.hazelcast.cache.CacheNotExistsException;
+import com.hazelcast.cache.impl.CacheEventType;
+import com.hazelcast.cache.impl.CacheService;
+import com.hazelcast.config.CacheConfig;
+import com.hazelcast.config.EventJournalConfig;
+import com.hazelcast.config.InMemoryFormat;
+import com.hazelcast.config.RingbufferConfig;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.DataType;
+import com.hazelcast.ringbuffer.impl.ReadResultSetImpl;
+import com.hazelcast.ringbuffer.impl.RingbufferContainer;
+import com.hazelcast.ringbuffer.impl.RingbufferService;
+import com.hazelcast.ringbuffer.impl.RingbufferWaitNotifyKey;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.ObjectNamespace;
+import com.hazelcast.spi.WaitNotifyKey;
+import com.hazelcast.spi.impl.NodeEngineImpl;
+import com.hazelcast.spi.impl.operationparker.OperationParker;
+
+import static com.hazelcast.cache.impl.CacheEventType.CREATED;
+import static com.hazelcast.cache.impl.CacheEventType.EVICTED;
+import static com.hazelcast.cache.impl.CacheEventType.EXPIRED;
+import static com.hazelcast.cache.impl.CacheEventType.REMOVED;
+import static com.hazelcast.cache.impl.CacheEventType.UPDATED;
+
+
+/**
+ * The cache event journal implementation based on the {@link com.hazelcast.ringbuffer.Ringbuffer}.
+ * It will add all journal events into a {@link RingbufferContainer} with the provided namespace
+ * and partition ID and allows checking if the cache has a configured event journal.
+ * Adapts the {@link EventJournalConfig} to the {@link RingbufferConfig} when creating the ringbuffer.
+ */
+public class RingbufferCacheEventJournalImpl implements CacheEventJournal {
+    private final NodeEngineImpl nodeEngine;
+
+    public RingbufferCacheEventJournalImpl(NodeEngine engine) {
+        this.nodeEngine = (NodeEngineImpl) engine;
+    }
+
+    @Override
+    public void writeUpdateEvent(ObjectNamespace namespace, int partitionId, Data key, Object oldValue, Object newValue) {
+        addToEventRingbuffer(namespace, partitionId, UPDATED, key, oldValue, newValue);
+    }
+
+    @Override
+    public void writeCreatedEvent(ObjectNamespace namespace, int partitionId, Data key, Object value) {
+        addToEventRingbuffer(namespace, partitionId, CREATED, key, null, value);
+    }
+
+    @Override
+    public void writeRemoveEvent(ObjectNamespace namespace, int partitionId, Data key, Object value) {
+        addToEventRingbuffer(namespace, partitionId, REMOVED, key, value, null);
+    }
+
+    @Override
+    public void writeEvictEvent(ObjectNamespace namespace, int partitionId, Data key, Object value) {
```

Comment:
It can be better to note that different entries can be evicted on primary and on backup as a result of the same call. So for example when `cache#put` returns, its possible to see that primary evicted entry-A but backup evicted entry-B. This leads to see different journals between replicas.